drivers/hmis/app/graphql/concerns/graphql_permission_checker.rb

```diff
@@ -8,6 +8,10 @@ def self.current_permission_for_context?(context, permission:, entity:)
     current_user = context[:current_user]
     return false unless current_user&.present?
 
+    if entity.data_source_id
```

Comment:
Will these entities always have `data_source_id`? I wasn't sure, but we can remove the guard if they always will. I was thinking since this is used on the base graphql object it won't always be present.

Reply:
Usually but not always, I don't think. I think this check needs to go after (or in) the `entity_access_loader_factory` which is responsible for basically translating the "entity" into the correct access Loader (Project, Organization, Client, or Data Source). _Most_ (or all) of the time, the entity that we pass does have a data_source_id. But there is nowhere that enforces that.

This will throw a `NoMethodError` if there is anywhere that we pass an entity without data_source_id. So at minimum lets roll that back or add a check for `respond_to?(:data_source_id)` so we don't get a new bug.


For more detail on the entity loader..it supports passing any record that has a project association (resolve_through_project), so it SHOULD work on a record associated to project, even if the record doesn't have data_source_id. I'm not sure if we have a usage like that right now, but I don't think we should prevent it. https://github.com/greenriver/hmis-warehouse/blob/7690e9f42e5083fb676468211443eb7acee1eb41/drivers/hmis/app/models/hmis/entity_access_loader_factory.rb#L70-L80 


So lets look at this more closely, I'd be happy to pair on it!

Reply:
Thanks, good point, and sorry for merging this prematurely! I opened https://github.com/greenriver/hmis-warehouse/pull/4788 adding `respond_to?(:data_source_id)` to prevent this bug. but let me take a minute to grok the larger change of moving this into the entity access loader factory...

Reply:
thanks! ok lets discuss when you're ready!